File: Rust/instance.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::mem;
use std::path::Path;

#[derive(Copy, Clone)]
pub struct Coords {
    x: f64,
    y: f64,
}

#[derive(Copy, Clone)]
pub struct Edge {
    pub v: usize,
    pub u: usize,
}

type Dist = fn(Coords, Coords) -> i32;

pub fn dist_one(_v: Coords, _u: Coords) -> i32 {
    return 1;
}

pub fn dist_euc2d(v: Coords, u: Coords) -> i32 {
    let dx = v.x - u.x;
    let dy = v.y - u.y;
    return ((dx * dx + dy * dy).sqrt() + 0.5) as i32;
}

pub fn dist_att(v: Coords, u: Coords) -> i32 {
    let xd = v.x - u.x;
    let yd = v.y - u.y;
    let rij = ((xd * xd + yd * yd) / 10.0).sqrt();
    let tij = (rij + 0.5) as i32;
    return if (tij as f64) < rij { tij + 1 } else { tij };
}

pub fn dist_ceil2d(v: Coords, u: Coords) -> i32 {
    let dx = v.x - u.x;
    let dy = v.y - u.y;
    return (dx * dx + dy * dy).sqrt().ceil() as i32;
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

#[derive(Clone)]
pub struct Instance {
    name: String,
    dist_type: String,
    coords: Vec<Coords>,
    d: Dist,
}

impl Instance {
    pub fn from_file(file_path: &str) -> Instance {
        let mut inst = Instance {
            name: "".to_string(),
            dist_type: "".to_string(),
            coords: vec![],
            d: dist_one,
        };
        if let Ok(lines) = read_lines(file_path) {
            let mut coord_to_read = 0;
            let mut dimension = -1;
            for line in lines {
                let l = line.unwrap();
                if coord_to_read > 0 {
                    coord_to_read -= 1;
                    //println!("{}", l);
                    let splitted: Vec<&str> = l.split_whitespace().collect();
                    let id: usize = splitted[0].parse::<usize>().unwrap();
                    let x: f64 = splitted[1].parse::<f64>().unwrap();
                    let y: f64 = splitted[2].parse::<f64>().unwrap();
                    inst.coords[id - 1] = Coords { x: x, y: y };
                } else {
                    let splitted: Vec<&str> = l.split(":").collect();
                    let keyword: &str = splitted[0];
                    match keyword.trim() {
                        "NAME" => {
                            let value: &str = &splitted[1..].join("");
                            inst.name = value.to_string();
                            println!("Name: {}", inst.name);
                        }
                        "DIMENSION" => {
                            let value: &str = &splitted[1..].join("");
                            dimension = value.trim().parse::<i32>().unwrap();
                            inst.coords
                                .resize(dimension as usize, Coords { x: 0.0, y: 0.0 });
                            println!("Dimension: {}", value);
                        }
                        "EDGE_WEIGHT_TYPE" => {
                            let value: &str = &splitted[1..].join("");
                            println!("Distance function: {}", value);
                            match value.trim() {
                                "EUC_2D" => inst.d = dist_euc2d,
                                "CEIL_2D" => inst.d = dist_ceil2d,
                                "ATT" => inst.d = dist_att,
                                _ => eprint!("Distance function not supported\n"),
                            }
                        }
                        "NODE_COORD_TYPE" => {
                            let value: &str = &splitted[1..].join("");
                            println!("Coordinates type: {}", value);
                        }
                        "NODE_COORD_SECTION" => {
                            assert_ne!(dimension, -1, "Dimension not read yet!");
                            coord_to_read = dimension;
                            println!("Coords section:");
                        }
                        "DEPOT_SECTION" => {
                            println!("Depot section:");
                        }
                        _ => {
                            eprint!("Keyword not recognized! {}\n", keyword)
                        }
                    }
                }
            }
        }
        return inst;
    }

    pub fn get_name(&self) -> &str {
        return self.name.as_ref();
    }
    pub fn get_dist_type(&self) -> &str {
        return self.dist_type.as_ref();
    }

    pub fn size(&self) -> usize {
        return self.coords.len();
    }

    pub fn get_x(&self, i: u32) -> f64 {
        return self.coords[i as usize].x;
    }

    pub fn get_y(&self, i: u32) -> f64 {
        return self.coords[i as usize].y;
    }

    pub fn dist(&self, v: u32, u: u32) -> i32 {
        return (self.d)(self.coords[v as usize], self.coords[u as usize]);
    }
}

File: Zig/instance.zig
const std = @import("std");
const assert = std.debug.assert;
const Allocator = std.mem.Allocator;

pub fn print(comptime fmt: []const u8, args: anytype) void {
    const stdout = std.io.getStdOut().writer();
    nosuspend stdout.print(fmt, args) catch return;
}

const Coord = struct { x: f32, y: f32 };

const Dist = *const fn (v: Coord, u: Coord) i32;

pub const Instance = struct {
    name: []const u8,
    dist_type: []const u8,
    coords: []Coord,
    d: Dist,

    pub fn size(self: Instance) usize {
        return self.coords.len;
    }

    pub fn dist(self: Instance, v: u32, u: u32) i32 {
        assert(v < self.size() and u < self.size());
        return self.d(self.coords[v], self.coords[u]);
    }
};

pub fn dist_one(_: Coord, _: Coord) i32 {
    return 1;
}

pub fn dist_euc2d(v: Coord, u: Coord) i32 {
    const dx: f64 = v.x - u.x;
    const dy: f64 = v.y - u.y;
    return @intFromFloat(@sqrt(dx * dx + dy * dy) + 0.5);
}

pub fn dist_att(v: Coord, u: Coord) i32 {
    const xd: f64 = v.x - u.x;
    const yd: f64 = v.y - u.y;
    const rij: f64 = @sqrt((xd * xd + yd * yd) / 10.0);
    const tij: i32 = @as(i32, @intFromFloat(rij + 0.5));
    if (tij < @as(i32, @intFromFloat(rij)))
        return tij + 1;
    return tij;
}

pub fn dist_ceil2d(v: Coord, u: Coord) i32 {
    const dx: f64 = v.x - u.x;
    const dy: f64 = v.y - u.y;
    return @as(i32, @intFromFloat(@ceil(@sqrt(dx * dx + dy * dy))));
}

const TsplibKeys = enum {
    NAME,
    DIMENSION,
    EDGE_WEIGHT_TYPE,
    NODE_COORD_TYPE,
    NODE_COORD_SECTION,
    DEPOT_SECTION,
    UNDEFINED,
};

const DistKeys = enum {
    ONE,
    EUC_2D,
    ATT,
    CEIL_2D,
    UNDEFINED,
};

pub fn read_inst_file(path: []const u8, allocator: Allocator) !Instance {
    const inst_file = try std.fs.cwd().openFile(path, .{});
    defer inst_file.close();

    var buf_reader = std.io.bufferedReader(inst_file.reader());
    var in_stream = buf_reader.reader();

    var buf: [1024]u8 = undefined;
    var inst: Instance = undefined;
    var dim: usize = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        var tokens = std.mem.split(u8, line, " ");
        const tk = tokens.next() orelse continue; //skip empty lines

        const okeyw = std.meta.stringToEnum(TsplibKeys, tk);
        if (okeyw) |k| {
            _ = tokens.next(); // colon
            switch (k) {
                .NAME => {
                    inst.name = try allocator.dupe(u8, tokens.rest());
                    print("NAME: {s}\n", .{inst.name});
                },
                .DIMENSION => {
                    dim = try std.fmt.parseInt(usize, tokens.rest(), 10);
                    print("DIMENSION: {s}\n", .{tokens.rest()});
                },
                .EDGE_WEIGHT_TYPE => {
                    inst.dist_type = try allocator.dupe(u8, tokens.rest());
                    print("EDGE_WEIGHT_TYPE: {s}\n", .{inst.dist_type});
                    const odist = std.meta.stringToEnum(DistKeys, inst.dist_type);
                    if (odist) |d| {
                        inst.d = switch (d) {
                            .ONE => dist_one,
                            .EUC_2D => dist_euc2d,
                            .ATT => dist_att,
                            .CEIL_2D => dist_ceil2d,
                            else => undefined,
                        };
                    }
                },
                .NODE_COORD_SECTION => {
                    print("NODE_COORD_SECTION:\n", .{});
                    assert(dim > 0);
                    inst.coords = try allocator.alloc(Coord, dim);
                    for (0..dim) |i| {
                        const c_line = (try in_stream.readUntilDelimiterOrEof(&buf, '\n')).?;
                        var c_tokens = std.mem.splitScalar(u8, c_line, ' ');
                        const id = try std.fmt.parseInt(usize, c_tokens.next().?, 10);
                        const x = try std.fmt.parseFloat(f32, c_tokens.next().?);
                        const y = try std.fmt.parseFloat(f32, c_tokens.next().?);

                        inst.coords[id - 1] = Coord{ .x = x, .y = y };
                        assert(id - 1 == i);
                    }
                },
                .DEPOT_SECTION => {
                    print("DEPOT_SECTION:\n", .{});
                    const id = try std.fmt.parseInt(i32, tokens.next().?, 10); // id
                    assert(id == 1);
                },
                else => print("Keyword not recognized: {s}\n", .{tk}),
            }
        } else {
            print("Keyword not recognized: {s}\n", .{tk});
        }
    }

    return inst;
}
